python/cudf/cudf/core/indexing_utils.py

```diff
@@ -241,3 +267,356 @@ def parse_row_iloc_indexer(key: Any, n: int) -> IndexingSpec:
+) -> tuple[Any, ...]:
+    """
+    Destructure a potentially tuple-typed key into row and column indexers
+
+    Tuple arguments to loc indexing are treated specially. They are
+    picked apart into indexers for the row and column. If the number
+    of entries is less than the number of modes of the frame, missing
+    entries are slice-expanded.
+
+    If the user-provided key is not a tuple, it is treated as if it
+    were a singleton tuple, and then slice-expanded.
+
+    Once this destructuring has occurred, any entries that are
+    callables are then called with the indexed frame. This should
+    return a valid indexing object for the rows (respectively
+    columns), namely one of:
+
+    - A boolean mask of the same length as the frame in the given
+      dimension
+    - A scalar label looked up in the index
+    - A scalar integer that indexes the frame
+    - An array-like of labels looked up in the index
+    - A slice of the index
+    - For multiindices, a tuple of per level indexers
+
+    Slice-based indexing is on the closed interval [start, end], rather
+    than the semi-open interval [start, end)
+
+    Parameters
+    ----------
+    key
+        The key to destructure
+    frame
+        DataFrame or Series to provide context
+
+    Returns
+    -------
+    tuple of indexers with length equal to the dimension of the frame
+
+    Raises
+    ------
+    IndexError
+        If there are too many indexers.
+    """
+    return expand_key(key, frame)
+
+
+def destructure_dataframe_loc_indexer(
+    key: Any, frame: cudf.DataFrame
+) -> Tuple[Any, Tuple[bool, ColumnAccessor]]:
+    """Destructure an index key for DataFrame loc getitem.
+
+    Parameters
+    ----------
+    key
+        Key to destructure
+    frame
+        DataFrame to provide context context
+
+    Returns
+    -------
+    tuple
+        2-tuple of a key for the rows and tuple of
+        (column_index_is_scalar, ColumnAccessor) for the columns
+
+    Raises
+    ------
+    TypeError
+        If the column indexer is invalid
+    IndexError
+        If the provided key does not destructure correctly
+    NotImplementedError
+        If the requested column indexer repeats columns
+    """
+
+    def is_scalar(name, ca):
+        try:
+            return name in ca
+        except TypeError:
+            return False
+
+    return destructure_dataframe_indexer(
+        key, frame, destructure_loc_key, is_scalar, "select_by_label"
+    )
+
+
+def destructure_series_loc_indexer(key: Any, frame: cudf.Series) -> Any:
+    """Destructure an index key for Series loc getitem.
+
+    Parameters
+    ----------
+    key
+        Key to destructure
+    frame
+        Series for unpacking context
+
+    Returns
+    -------
+    Single key that will index the rows
+    """
+    (rows,) = destructure_loc_key(key, frame)
+    return rows
+
+
+def ordered_find(needles: "ColumnBase", haystack: "ColumnBase") -> GatherMap:
```

Comment:
This is basically just `DataFrame.merge(sort=True)` but I don't _really_ want to pay the price of making dataframes and all of the merge inspection machinery.